### DISCUSSION TITLE
What's the best/recommended way to mock auto-curried functions in tests?

### DISCUSSION THREAD

**Discussion by nlucero** [ramda/ramda#3204] [Q&A]
I'm trying to write some tests with `Jest` for a piece code I wrote that consumes an auto-curried function (i.e. a function curried with [`curry`](https://ramdajs.com/docs/#curry)) and having some troubles mocking it.

A simplified version of my code:

```js
// service/adder
export const add = curry((a, b) => a + b)
```

```js
// service/transformer
import { add } from './adder'

export const transform = add(11)
```

Using auto-mock is not possible for obvious reasons: The `add` function is mocked with a simple `jest.fn` that returns `undefined` when invoked, so `add(11)` is `undefined` and any call to `transform` will fail:

```js
// service/transformer.test.js
import { transform } from './transformer'

jest.mock('./service/adder')

it('should transform the supplied value', () => {
  const result = transform(123) // Boom!
})
```

I could write a manual curried mock that returns an instance of `jest.fn` after each argument is supplied but it seems tedious/verbose, and also doing assertions like `expect(add).toHaveBeenCalledWith(11, 123)` become much harder to do.

Are there any better/recommended ways to do this? :thinking:

**Comment by nlucero** [comment]
Ended up doing this, which seems to work:

```js
// service/transformer.test.js
import { transform } from './transformer'

jest.mock('./service/adder', () => ({ add: jest.fn().mockReturnValue(jest.fn()) }))

it('should transform the supplied value', () => {
  add().mockReturnValueOnce(999)

  const result = transform(123)

  expect(result).toBe(999)
  expect(add()).toHaveBeenCalledWith(123)
})
```

This works fine in this simple scenario, but would quickly become tedious if the mocked function is partially applied in different ways in the same file :thinking:

**Reply by nfantone** [reply]
I think your approach is what I would call "recommended" - without the need for further testing utilities or helper functions, that is.

Another, arguably less neat, approach would be to initialize the mock _after_ requiring your `curry` dependant module, if you can afford it.

```js
'use strict'
const { add } = require('./adder')
jest.mock('./adder')

test('should call curried function', () => {
  const doSum = jest.fn((a, b) => a + b)
  add.mockImplementationOnce(a => b => doSum(a, b))

  // transform.js required here so implementation picks up the mock above
  const { transform } = require('./transform')

  expect(transform(1)).toBe(12)

  expect(doSum).toHaveBeenCalledOnce()
  expect(doSum).toHaveBeenCalledWith(11, 1)
})
```

This way, at least, you give control over the way the curried function behaves to each individual test - somewhat addressing the concerns you raised in your last point.
